ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -232,22 +258,53 @@ void dropBufferForAppend() {
    */
   @Override
   public void readInto(ByteBuffer buffer, long position) throws IOException {
+    int size = buffer.remaining();
+    validateReadRange(position, size);
+    long bytesRead = 0;
+    while (bytesRead < size) {
+      bytesRead += fileChannel.read(buffer, position + bytesRead);
+    }
+  }
+
+  /**
+   * @param byteArray The buffer into which the data needs to be written.
+   * @param position The source's position to start the read.
+   * @param length The length of data in bytes to read.
+   * @throws IOException if data could not be written to the file because of I/O errors.
+   * @throws IndexOutOfBoundsException if {@code position} < header size or >= {@link #sizeInBytes()} or if
+   * {@code buffer} size is greater than the data available for read.
+   */
+  void readIntoDirectly(byte[] byteArray, long position, int length) throws IOException {
+    validateReadRange(position, length);
+    if (file.length() < position + length) {
```

Comment:
isn't this check already done in `validateReadRange()`?

Reply:
fixed.